PR: Merging gps_main_branch to main_branch

ganeshram93 opened the pull request:
Pull Request - GPS module 
1. Interfaced GPS hardware and added the GPS module
2. Handled cases when there is Inactive GPS mode and if GPS hardware stops buffering data - NA will be sent to BT module

When GPS module is Connected:
![image](https://user-images.githubusercontent.com/19392049/109585021-88f08500-7ab7-11eb-97e5-0501ef7f0c35.png)

When GPS module is Disconnected/Not working/Inactive/No GPRMC data in buffer:
![image](https://user-images.githubusercontent.com/19392049/109585218-ec7ab280-7ab7-11eb-81bd-26a126a57f8b.png)

ganeshram93 committed:
1. Interfaced GPS hardware and added the GPS module
2. Handled cases when there is Inactive GPS mode and if GPS hardware stops buffering data - NA will be sent to BT module

coder137 commented on HelloWorld_ST_Sensors/user/gps/gps.cpp:
```diff
@@ -0,0 +1,152 @@
+            index++;
+        }
+        if (index == UART_BUF_LEN)
+        {
```
Why wait for it to be populated with UART_BUF_LEN?
Check for \n and then parse into latitude/longitude or call the registered_callback function

coder137 replied:
The second issue with this logic is that below you are taking the entire buffer of 400 characters and checking for only one received type.
Instead get GPS data per line, the above parsing logic would have much fewer if else lines

ganeshram93 replied:
Resolved in new PR.

coder137 commented on HelloWorld_ST_Sensors/user/gps/gps.cpp:
```diff
@@ -0,0 +1,152 @@
+        if (index == UART_BUF_LEN)
+        {
+            uint8_t gps_nmea_type = extract_gpgga_or_gprmc_line(curr_line_buffer);
+            if ((GPS_INVALID_NMEA != gps_nmea_type) && (gps_process_fn_ptr != NULL))
```
I can see that you have handled the case where gps_nmea_type is not valid for our purposes
And you are calling the callback function here. 

But there isn't any parsing logic for the gpgga / gprmc in the else clause.

coder137 replied:
Okay I realize you might be update the gpgga/gprmc in the `extract_gpgga_or_gprmc_line` function.
Please make the functions concise and limit their scope to one action

ganeshram93 replied:
Resolved in new PR.

coder137 commented on HelloWorld_ST_Sensors/user/gps/gps.cpp:
```diff
@@ -0,0 +1,152 @@
+        char valid_gps_string[100] = {0};
+        uint16_t counter = 0;
+        for (uint16_t i = 0; i < UART_BUF_LEN; i++)
+        {
```
As mentioned in the interrupt callback, you are iterating over the entire UART_BUF_LEN
But when you receive the first $ value you are parsing that and exiting
There can be multiple $ tokens received we should not need to miss any

ganeshram93 replied:
Resolved in new PR.

coder137 commented on HelloWorld_ST_Sensors/user/gps/gps.cpp:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    uint8_t gps_c::check_gprmc_gpgga_line(char *line_buff)
+    {
```
Please improve this function, `extract_gpgga_or_gprmc_line` and `gps_interrupt_cb`

Logic should be simple
- Receive one line
- Check if GGA/RMC

- If NO , call registered callback function
- If YES, parse and update lat lon and other data values in the class

ganeshram93 replied:
Resolved in new PR.

coder137 commented on HelloWorld_ST_Sensors/user/gps/gps.hpp:
```diff
@@ -0,0 +1,66 @@
+
+        typedef struct gps_data
+        {
+            char latitude[15];
```
latittude, longitude will be double values
lat dir and long dir can be enum values
validity can be a bool value

You can consider making different structs for GGA and RMC

ganeshram93 replied:
Resolved in new PR.

coder137 commented on HelloWorld_ST_Sensors/user/gps/gps.hpp:
```diff
@@ -0,0 +1,66 @@
+            GPS_GPRMC
+        };
+
+        enum GPS_NMEA_STRING_INDEX_DEFINITION
```
Will this definition be used by the user? If not put in private

ganeshram93 replied:
This should not be in private because other GPS modules can make use of the enum and struct.
In the new PR, we have moved the enums and struct to NS, outside the class.

coder137 commented on HelloWorld_ST_Sensors/user/gps/gps.hpp:
```diff
@@ -0,0 +1,66 @@
+            GPRMC_LONGITUDE_DIR_INDEX = 6,
+        };
+
+        char curr_line_buffer[UART_BUF_LEN];
```
Add getters for these variables. Users might accidentally modify this variable unnecessarily

ganeshram93 replied:
It's an enum type, so no need for getters.

coder137 commented on HelloWorld_ST_Sensors/user/gps/gps.hpp:
```diff
@@ -0,0 +1,66 @@
+#include "mbed.h"
+#include <string.h>
+
+namespace gps_ns
```
Improve naming

namespace gps
class gps device name. For example 
- gps::ublox_[part_number]
- gps::adafruit_[part_number]

ganeshram93 replied:
Resolved in new PR.

ganeshram93 committed:
1. Removed GPGGA related code
2. Added logic of reading '\n' and processing the line, rather than reading 400 characters at a given time and processing
3. Renamed class, namesspaces, and some function names

ganeshram93 reviewed (commented):
Updated PR after comments on first PR: Merging gps_main_branch to main_branch

ganeshram93 committed:
Updated main.cpp after resolving merge conflicts due to merging of bt branch in main

ganeshram93 committed:
1. Added comments for Interfaces for each sensor
2. Added default coordinate values (0.0, 0.0) that will be set when GPS module fails to function
